### PR TITLE
Poroelastic fracture, with broken permeability tensor

### PR DESCRIPTION
**Pull request by kmokstad**
This one replaces #3. Contains in addition a method for calculation of the anisotropic permeability of the cracked material, based on Miehe (2015), Part III. Please take a look at it.

The PoroElasticity class now needs a virtual version of evalPermeabilityMatrix in which the new method is used.

### CODE REVIEW

**Comment by akva2** [comment]
jeg er imot alt dette.

File: PoroFracture.C [hunk 1]
```diff
@@ -164,3 +188,51 @@ bool PoroFracture::evalElasticityMatrices (ElmMats& elMat, const Matrix&,
   // Evaluate tangent stiffness matrix and internal forces
   return fracEl->evalInt(elMat,fe,X);
 }
+
+
+/*!
+  This method calculates the anisotropic peremeability for the broken solid
```
Comment (akva2) [inline PoroFracture.C]: permeabilty

File: PoroFracture.C [hunk 2]
```diff
@@ -164,3 +188,51 @@ bool PoroFracture::evalElasticityMatrices (ElmMats& elMat, const Matrix&,
...
+  based on a Poiseuielle flow approximation of the fluid flow in the crack.
```
Comment (akva2) [inline PoroFracture.C]: pouiseville

**Comment by akva2** [comment]
vennligst inkluder en fiks for
```
In file included from /var/lib/jenkins/jobs/IFEM-PR-builder/workspace/deps/IFEM-OpenFrac/FractureElasticityVoigt.C:14:
/var/lib/jenkins/jobs/IFEM-PR-builder/workspace/deps/IFEM-OpenFrac/FractureElasticityVoigt.h:14:9: warning: '_FRACTURE_ELASTICITY_VOIGT_H' is used as a header guard here, followed by #define of a different macro [-Wheader-guard]
#ifndef _FRACTURE_ELASTICITY_VOIGT_H
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/jenkins/jobs/IFEM-PR-builder/workspace/deps/IFEM-OpenFrac/FractureElasticityVoigt.h:15:9: note: '_FRACTURE_ELASTICITY_VOIGH_H' is defined here; did you mean '_FRACTURE_ELASTICITY_VOIGT_H'?
#define _FRACTURE_ELASTICITY_VOIGH_H
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
        _FRACTURE_ELASTICITY_VOIGT_H
```

vil ikke røre brancher du holder som gissel..

**Comment by kmokstad** [comment]
jenkins build this please

File: PoroFracture.C [hunk 3]
```diff
@@ -0,0 +1,344 @@
+// $Id$
+//==============================================================================
+//!
+//! \file PoroFracture.C
+//!
+//! \date Apr 15 2016
+//!
+//! \author Knut Morten Okstad / SINTEF
+//!
+//! \brief Integrand implementations for elasticity problems with fracture.
```
Comment (akva2) [inline PoroFracture.C]: poro

File: PoroFracture.C [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+#include "FiniteElement.h"
+#include "Tensor.h"
+#include "Vec3Oper.h"
+#include "Utilities.h"
+#include "IFEM.h"
+#include "tinyxml.h"
+
+
+/*!
+  \brief Class representing the integrand of elasticity problems with fracture.
```
Comment (akva2) [inline PoroFracture.C]: poro

File: PoroFracture.h [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  //! \param[in] fld Which field set to consider (1=primary,2=secondary)
+  virtual size_t getNoFields(int fld) const;
+
+  //! \brief Returns the name of a secondary solution field component.
+  //! \param[in] i Field component index
+  //! \param[in] prefix Name prefix for all components
+  virtual std::string getField2Name(size_t i, const char* prefix) const;
+
+protected:
+  //! \brief Computes the elasticity matrices for a quadrature point.
```
Comment (akva2) [inline PoroFracture.h]: at?

File: PoroFracture.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  //! \brief Initializes current element for numerical integration.
+  //! \param[in] MNPC Matrix of nodal point correspondance for current element
+  //! \param elmInt The local integral object for current element
+  virtual bool initElement(const std::vector<int>& MNPC, LocalIntegral& elmInt);
+  //! \brief Initializes current element for numerical integration.
+  //! \param[in] MNPC Matrix of nodal point correspondance for current element
+  //! \param[in] elem_sizes Size of each basis on the element
+  //! \param[in] basis_sizes Size of each basis on the patch level
+  //! \param elmInt The local integral object for current element
+  virtual bool initElement(const std::vector<int>& MNPC,
```
Comment (akva2) [inline PoroFracture.h]: som sagt i møtet; initElement har en mixed-overload men det er ingen evalIntMx. men jeg ser nå at det bruker default overload'en (kall non-mixed) så det bør være okay.

File: PoroFracture.C (hunk 3 above)
Comment (kmokstad) [reply PoroFracture.C]: Fixed

File: PoroFracture.C (hunk 4 above)
Comment (kmokstad) [reply PoroFracture.C]: Fixed

File: PoroFracture.h (hunk 5 above)
Comment (kmokstad) [reply PoroFracture.h]: yes, fixed

File: PoroFracture.h (hunk 6 above)
Comment (kmokstad) [reply PoroFracture.h]: Ja, jeg mener å huske at -mixed var operativt når Eivind drev og prøvde for noen år siden.
Jeg har kun en input-file liggende Miehe71.xinp i Test-katalogen, men intet test-opsett forøvring.

File: SIMExplPhaseField.C [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+#include "DataExporter.h"
+#include "Functions.h"
+#include "Utilities.h"
+#include "IFEM.h"
+#include "tinyxml.h"
+
+
+SIMExplPhaseField::SIMExplPhaseField (SIMoutput* gridOwner)
+{
+  myHeading = "Explcit phase field";
```
Comment (akva2) [inline SIMExplPhaseField.C]: explicit

File: SIMExplPhaseField.C [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+{
+  this->registerField("phasefield",phaseField);
+  return true;
+}
+
+
+bool SIMExplPhaseField::parse (const TiXmlElement* elem)
+{
+  const char* value = utl::getValue(elem,"phasefield");
+  if (!value) return false;
```
Comment (akva2) [inline SIMExplPhaseField.C]: this doesn't seem right. we are passed all subtags of <simulation>, not all will match phasefield and returning false here will halt parsing.

**Comment by akva2** [comment]
the explicit phase field construct is rather problematic for mixed. it uses the model from the elasticity problem, which is mixed. while i can fix the projection to only project onto basis 1, it's not fixable when extracting the patch dependencies. as we are in a SIMdependency scope there, i do not have access to the additional MADOF required for a single component on first basis (which has to sit in the elasticity SIM). i see no clean way around setting up a model in the phase sim itself.

File: SIMExplPhaseField.C (hunk 7 above)
Comment (kmokstad) [reply SIMExplPhaseField.C]: Fixed.

File: SIMExplPhaseField.C (hunk 8 above)
Comment (kmokstad) [reply SIMExplPhaseField.C]: Indeed wrong. Should here forward to parent-class method for all other tags. Don't know when that vanished but fixed now. Probably something went wrong during a rebase at one point.

**Comment by kmokstad** [comment]
Have to think about that. I see the program crashes now running with -explcrack and -mixed, while it runs (but diverges) without the -mixed option. I guess these two options have never been combined before, unless @TheBB remembers doing so back in 2016?

**Comment by akva2** [comment]
fixed explicit phase field plus mixed. it's not entirely pretty as this is not using the established mechanisms to handle this (fields). you need https://github.com/akva2/IFEM-OpenFrac/tree/fix_poro_expl and https://github.com/akva2/IFEM/tree/fix_expl_mixed

**Comment by kmokstad** [comment]
This branch has now been rebased and the items from https://github.com/akva2/IFEM-OpenFrac/tree/fix_poro_expl is included in a slightly modified version in the two last commits.
Also there was a bug when running with -poro -mixed and not using the explicit phase field. Then the grid-sharing between the poroelasticity and phase field solvers have to be switched off. So with this the two solvers (explicit and unknown phase field) behaves somewhat similar.

Problably something needs to be done when going to adaptive refinement and mixed when the grids no longer shared, but den tid den sorg...

Think it is time to merge this branch now, before continuing on this track (provided all checks passes)

**Comment by akva2** [comment]
jenkins build this ifem=286 ifem-poroelasticity=27 valgrind please

**Comment by akva2** [comment]
jenkins build this ifem=286 ifem-poroelasticity=27 valgrind please

**Comment by akva2** [comment]
jenkins build this ifem=286 ifem-poroelasticity=27 valgrind please

**Comment by kmokstad** [comment]
With the last change here (that is, we specify an additional MADOF also when phase field solver is in the loop and not only with explicit field), and yes, the phase-field solver has to share grid with the poro-elastic solver, only specyfying zero dofs pr node for second basis and 1 dofs per node for the first, then with the OPM/IFEM#289 fix the Miehe71.xinp case with -poro -mixed runs through without valgrind issues or divergence. But the solution is totally wrong still (look atht the phase field in GLview). But that is probably due to other things. I think the use of Fields for the phase-field dependency when mixed should be tried out next, to see if that gives a better solution. In that case the current implementation is still buggy. If not, we have a problem with the formulation as such,

**Comment by akva2** [comment]
If I
- adjust ngauss to 3
- adjust Gc to values more aligned to the return energies

it runs and looks sane. of course, i'm not really sure what's supposed to happen in this case though.

**Comment by kmokstad** [comment]
jenkins build this please

**Comment by kmokstad** [comment]
jenkins build this please

**Comment by kmokstad** [comment]
jenkins build this please

**Comment by kmokstad** [comment]
The last comments above (from Sep 2 and Sep 7) applies to commits which now are placed in #37 instead.

**Pull request merged**